src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -79,29 +267,50 @@ extern "C" SortHandle* GetSortHandle(const char* lpLocaleName)
 extern "C" void CloseSortHandle(SortHandle* pSortHandle)
 {
     ucol_close(pSortHandle->regular);
-    ucol_close(pSortHandle->ignoreCase);
-
     pSortHandle->regular = nullptr;
-    pSortHandle->ignoreCase = nullptr;
+
+    TCollatorMap::iterator it;
+    for (it = pSortHandle->collatorsPerOption.begin(); it != pSortHandle->collatorsPerOption.end(); it++)
+    {
+        ucol_close(it->second);
+    }
 
     delete pSortHandle;
 }
 
-const UCollator* GetCollatorFromSortHandle(const SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
+const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
 {
-    if ((options & CompareOptionsIgnoreCase) == CompareOptionsIgnoreCase)
+    UCollator* pCollator;
+    if (options == 0)
+    {
+        pCollator = pSortHandle->regular;
+    }
+    else
     {
-        return pSortHandle->ignoreCase;
+        pthread_mutex_lock(&pSortHandle->collatorsLockObject);
+
+        TCollatorMap::iterator entry = pSortHandle->collatorsPerOption.find(options);
+        if (entry == pSortHandle->collatorsPerOption.end())
+        {
+            pCollator = CloneCollatorWithOptions(pSortHandle->regular, options, pErr);
```

Comment:
There is a potential for performance impact with this implementation.  If an application creates and throws away a lot of CompareInfo instances (i.e. CultureInfo instances), we will need to create new SortHandle instances for each CompareInfo instance.  Thus with a new SortHandle, the collator table will be empty, so we will need to build it again on demand.  Creating a collator from custom rules isn't a "free" operation, from what I'm told. However, I haven't yet measured the perf impact.
Note that we need to create a custom rule collator in the normal case of just using CurrentCultureIgnoreCase, because in that scenario IgnoreCase is true, and IgnoreWidth and IgnoreKanaType are false.  Since ICU supports casing, width and kana differences with the single "strength" setting, we use custom rules to ensure we are still differentiating between width and kana types, as requested by the caller.
If we find a lot of applications creating new CultureInfos, and need to fix this, we could cache the SortHandle instances across multiple CompareInfo objects.
Another option would be to not use custom rules, or relax certain times when we do (like if just IgnoreCase is set, and nothing else).  This wouldn't be functionally correct because it would also be ignoring width and kana character differences and the behavior would differ from Windows, but it might suffice for most applications.
Another option would be to not use ICU to do string comparisons, which doesn't seem logical.
A last option would be to change the API, which I'm sure isn't really an option. But just putting it out there.